Objects/obmalloc.c

```diff
@@ -1141,8 +1141,29 @@ free_work_item(uintptr_t ptr, delayed_dealloc_cb cb, void *state)
     }
 }
 
+#ifdef Py_GIL_DISABLED
+static int
+should_advance_qsbr(_PyThreadStateImpl *tstate, size_t size)
+{
+    // If the deferred memory exceeds 1 MiB, we force an advance in the
+    // shared QSBR sequence number to limit excess memory usage.
+    static const size_t QSBR_DEFERRED_LIMIT = 1024 * 1024;
+    if (size > QSBR_DEFERRED_LIMIT) {
+        tstate->qsbr->memory_deferred = 0;
+        return 1;
+    }
+
+    tstate->qsbr->memory_deferred += size;
+    if (tstate->qsbr->memory_deferred > QSBR_DEFERRED_LIMIT) {
+        tstate->qsbr->memory_deferred = 0;
+        return 1;
+    }
+    return 0;
+}
```

Comment:
Do we need early return here?
It looks like it will be same eventually.
```suggestion
should_advance_qsbr(_PyThreadStateImpl *tstate, size_t size)
{
    // If the deferred memory exceeds 1 MiB, we force an advance in the
    // shared QSBR sequence number to limit excess memory usage.
    static const size_t QSBR_DEFERRED_LIMIT = 1024 * 1024;
    tstate->qsbr->memory_deferred += size;
    if (tstate->qsbr->memory_deferred > QSBR_DEFERRED_LIMIT) {
        tstate->qsbr->memory_deferred = 0;
        return 1;
    }
    return 0;
}
```